### ISSUE TITLE
nsfs: Not able to create mutilevel directory if object of the same name exist

### ISSUE THREAD

**Issue by nishant16** [noobaa/noobaa-core#6412]
### Environment info
- NooBaa Version:  5.7
- Platform: OpenShift 4.6

### Actual behavior
1. creation of object inside directory of bucket:
```
[noob@noobaa-endpoint-6d87469b9f-qx2gd nishant-36uth8dq7d4fe7jdsdpkk67]$ ls
foo
[noob@noobaa-endpoint-6d87469b9f-qx2gd nishant-36uth8dq7d4fe7jdsdpkk67]$ cd foo/
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ ls
bar
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ ls
bar
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ cd bar
bash: cd: bar: Not a directory
```
Creating file inside foo:
```
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ cd ..
[noob@noobaa-endpoint-6d87469b9f-qx2gd nishant-36uth8dq7d4fe7jdsdpkk67]$ ls -lrt
total 1
drwxr-xr-x. 2 noob root 4096 Mar 16 15:25 foo
[noob@noobaa-endpoint-6d87469b9f-qx2gd nishant-36uth8dq7d4fe7jdsdpkk67]$ cd foo/
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ ls -lrt
total 1
-rw-r--r--. 1 noob root 7 Mar 16 15:25 bar
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ file bar 
bash: file: command not found
[noob@noobaa-endpoint-6d87469b9f-qx2gd foo]$ mkdir bar
mkdir: cannot create directory 'bar': File exists
```
### Expected behavior
1. mkdir bar should work inside foo/

### Steps to reproduce
1.  s3 mb s3://bucket-name
2.  echo test | s3 cp - s3://bucket-name/foo/bar
3.  echo test | s3 cp - s3://bucket-name/foo/bar/zoo

**Comment by guymguym** [comment]
Hi @nishant16 
CC @troppens @motorman-ibm

There is a bit of a nuance here.
If you change the first operation to use the key `foo/bar/` - meaning with a trailing slash in the name, we will `mkdir foo/bar` as a directory to begin with, and then store the content of the object in special `foo/bar/.folder` filename inside (the name is configurable).

However in your test case you start by creating the key `foo/bar` as a file and after that to use it as a directory.
In that case, it is possible to transition the existing object in `foo/bar` to a dir and relocate the existing file to `foo/bar/.folder`. But the question is if this is really desired in terms of cross protocol support.

Another related test case is to first create `foo/bar/zoo` and then try to create `foo/bar` which will currently fail with EISDIR.

I think the design should consider if the notion of a directory can leak to the S3 protocol in order to keep cross protocol support easier, or we prefer to make all transitions between files/dirs in order to hide it from the S3 protocol. I personally think that cross protocol would be a more relevant case here because the application can still use the `foo/bar/` name (with trailing slash) as a valid object key if desired. But it would be good to get your thoughts on this.


See code paths here: 
https://github.com/noobaa/noobaa-core/blob/f1ec8ef7dd7ce3f09264e1f0cf7ebcae0493ff76/src/sdk/namespace_fs.js#L688-L691
https://github.com/noobaa/noobaa-core/blob/f1ec8ef7dd7ce3f09264e1f0cf7ebcae0493ff76/src/sdk/namespace_fs.js#L698-L701
https://github.com/noobaa/noobaa-core/blob/f1ec8ef7dd7ce3f09264e1f0cf7ebcae0493ff76/config.js#L547

**Comment by troppens** [comment]
I agree that the multi-protocol data access use case is more relevant and I think that it is acceptable to disallow certain (to be defined) characters for object names. In this way a trailing `/` would trigger the creation of a directory in the underlying filesystem.

For `foo/bar/zoo` we could argue whether we want to create a nested sub directory tree (e.g. using `mkdir -p`) or whether we want to return an error. I would vote for the  `mkdir -p`, but I do not have a strong opinion here.

Just for reference - Other S3 on File implementations have similar restrictions, e.g.:
https://docs.min.io/docs/minio-server-limits-per-tenant.html  (scroll down to the bottom of this page)
